File: src/services/show-service/show.mapper.ts
import {IEpisode} from "./types/episode.type";
import {IShow} from "./types/show.type";
import {ISeason} from "./types/season.type";

export class ShowMapper {

    static map(show: any) : IShow {
        return {
            identifier: show.Identifier,
            title: show.Title,
            slug: show.Slug,
            year: show.Year,
            tmdb: show.TMDB,
            posterUrl: show.Poster,
            overview: show.Overview,
            genres: show.Genres.map((x: any) => {
                return {
                    identifier: x.Identifier,
                    name: x.Name,
                };
            }),
        };
    }

    static mapSeason(season: any) : ISeason {
        return {
            identifier: season.Identifier,
            title: season.Title,
            number: season.Number,
            show: this.map(season.Show),
        };
    }

    public static mapEpisode(episode: any) : IEpisode {
        return {
            identifier: episode.Identifier,
            title: episode.Title,
            number: episode.Number,
            season: this.mapSeason(episode.Season),
        };
    }
}

File: api/Trackster.Api/Features/Shows/ShowMapper.cs
using Trackster.Api.Data.Records;
using Trackster.Api.Features.Media.Types;
using Trackster.Api.Features.Shows.Types;

namespace Trackster.Api.Features.Shows;

public static class ShowMapper
{

    public static Show Map(ShowRecord show, List<GenreRecord> genres)
    {
        return new Show
        {
            Identifier = show.Identifier,
            Title = show.Title,
            Slug = show.Slug,
            Year = show.Year,
            TMDB = show.TMDB,
            Poster = show.Poster,
            Overview = show.Overview,
            Genres = genres.ConvertAll(x => new Genre
            {
                Identifier = x.Identifier,
                Name = x.Name
            })
        };
    }

    public static Season MapSeason(SeasonRecord season, List<GenreRecord> genres)
    {
        return new Season
        {
            Identifier = season.Identifier,
            Title = season.Title,
            Number = season.Number,
            Show = Map(season.Show, genres)
        };
    }
    
    public static Episode MapEpisode(EpisodeRecord episode, List<GenreRecord> genres)
    {
        return new Episode
        {
            Identifier = episode.Identifier,
            Title = episode.Title,
            Number = episode.Number,
            Season = MapSeason(episode.Season, genres),
        };
    }
}